sympy/integrals/transforms.py

```diff
@@ -1186,607 +1172,549 @@ def _laplace_build_rules(t, s):
+            prep_old = prep
+        ma = prep_f.match(t_dom)
         if ma:
-            debug('_laplace_apply_rules match:')
-            debug('      f:    %s'%(func,))
-            debug('      rule: %s o---o %s'%(t_dom, s_dom))
             try:
-                debug('      try   %s'%(check,))
                 c = check.xreplace(ma)
-                debug('      check %s -> %s'%(check, c))
-                if c==True:
-                    return _laplace_cr(k*s_dom.xreplace(ma),
-                                plane.xreplace(ma), S.true, **hints)
-            except Exception:
-                debug('_laplace_apply_rules did not match.')
-    if f.has(DiracDelta):
-        return None
-
-    prog_rules = [_laplace_rule_timescale, _laplace_rule_heaviside,
-                  _laplace_rule_exp, _laplace_rule_trig, _laplace_rule_diff]
-    for p_rule in prog_rules:
-        L = p_rule(func, t, s, doit=doit, **hints)
-        if L is not None:
-            noconds = hints.get('noconds', False)
-            if not noconds and type(L) is tuple:
-                r, p, c = L
-                return (k*r, p, c)
-            else:
-                return k*L
+            except TypeError:
+                # This may happen if the time function has imaginary
+                # numbers in it. Then we give up.
+                continue
+            if c==True:
+                debug('_laplace_apply_simple_rules match:')
+                debug('      f:     %s'%(f,))
+                debug('      rule:  %s o---o %s'%(t_dom, s_dom))
+                debug('      match: %s'%(ma, ))
+                return (s_dom.xreplace(ma), plane.xreplace(ma), c)
     return None
 
+def _laplace_transform(fn, t_, s_, simplify=True):
+    """
+    Front-end function of the Laplace transform. It tries to apply all known
+    rules recursively, and if everything else fails, it tries to integrate.
+    """
+    debug('[LT _l_t] (%s, %s, %s)'%(fn, t_, s_))
+
+    terms = Add.make_args(fn)
+    results = []
+    for ff in terms:
+        k, ft = ff.as_independent(t_, as_Add=False)
+        if (r := _laplace_apply_simple_rules(ft, t_, s_)) is not None:
+            pass
+        elif (r := _laplace_apply_prog_rules(ft, t_, s_)) is not None:
+            pass
+        elif (r := _laplace_expand(ft, t_, s_)) is not None:
+            pass
+        elif ft.has(AppliedUndef):
+            r = (LaplaceTransform(ft, t_, s_), S.NegativeInfinity, True)
+        elif (r := _laplace_transform_integration(ft, t_, s_,
+                                      simplify=simplify)) is not None:
+            pass
+        else:
+            r = None
+        if r is not None:
+            results.append((k*r[0], r[1], r[2]))
+        else:
+            results.append((k*LaplaceTransform(ft, t_, s_),
+                            S.NegativeInfinity, True))
+
+    r = []
+    p = []
+    c = []
+    for res in results:
+        r.append(res[0])
+        p.append(res[1])
+        c.append(res[2])
```

Comment:
Rather than setting `r = None` you could just set
```python
r = (LaplaceTransform(ft, t_, s_), S.NegativeInfinity, True)
```
Then there only needs to be one place to remember to multiply by `k`. We already had a bug from forgetting to multiply by `k` somewhere so it really is good to limit the number of places where we need to keep track of things like that.

Also note that we're packing everything into a list of tuples only to unpack that list again after. We could just directly populate `r, p, c` in the list above like:
```
(ri, pi, ci) = r
r.append(ri)
p.append(pi)
c.append(ci)
```
This also shows that it's confusing to use the same name `r` for two different things and that single letter variable names should generally be discouraged except in particular mathematical contexts.

The code that I wrote above already does this and uses clearer variable names.

Reply:
True, good idea. Will change it.